### ISSUE TITLE
Symlinked Directories Cannot be Deleted in Windows

### ISSUE THREAD

**Issue by SteveLowe** [chef/chef#4071]
Trying to delete a symlinked directory on Windows results in an EACCESS error.
It looks like this is the same issue as [CHEF-3960](https://tickets.opscode.com/browse/CHEF-3960).

I have tested Chef 12.5 on Windows Server 2008 R2.

With this resource:
```
directory 'C:/Temp/B' do
  action :create
end
link "C:/Temp/A" do
  to 'C:/Temp/B'
end
link "C:/Temp/A" do
  action :delete
end
```

When running Chef the following is logged:
```
[2015-10-19T12:57:48+01:00] INFO: Auto-discovered chef repository at C:/temp/Chef
[2015-10-19T12:57:48+01:00] INFO: Started chef-zero at chefzero://localhost:8889 with repository at C:/temp/Chef
  One version per cookbook

Starting Chef Client, version 12.5.1
[2015-10-19T12:57:53+01:00] INFO: *** Chef 12.5.1 ***
[2015-10-19T12:57:53+01:00] INFO: Chef-client pid: 1436
[2015-10-19T12:58:09+01:00] INFO: Setting the run_list to [#<Chef::RunList::RunListItem:0x4284bd0 @version=nil, @type=:r
ecipe, @name="test">] from CLI options
[2015-10-19T12:58:09+01:00] INFO: Run List is [recipe[test]]
[2015-10-19T12:58:09+01:00] INFO: Run List expands to [test]
[2015-10-19T12:58:09+01:00] INFO: Starting Chef Run for <hostname>
[2015-10-19T12:58:09+01:00] INFO: Running start handlers
[2015-10-19T12:58:09+01:00] INFO: Start handlers complete.
[2015-10-19T12:58:09+01:00] INFO: HTTP Request Returned 404 Not Found: Object not found:
resolving cookbooks for run list: ["test"]
[2015-10-19T12:58:09+01:00] INFO: Loading cookbooks [test@0.1.0]
Synchronizing Cookbooks:
  - test (0.1.0)
Compiling Cookbooks...
[2015-10-19T12:58:09+01:00] WARN: Cloning resource attributes for link[C:/Temp/A] from prior resource (CHEF-3694)
[2015-10-19T12:58:09+01:00] WARN: Previous link[C:/Temp/A]: c:/chef/local-mode-cache/cache/cookbooks/test/recipes/defaul
t.rb:10:in `from_file'
[2015-10-19T12:58:09+01:00] WARN: Current  link[C:/Temp/A]: c:/chef/local-mode-cache/cache/cookbooks/test/recipes/defaul
t.rb:13:in `from_file'
Converging 3 resources
Recipe: test::default
  * directory[C:/Temp/B] action create[2015-10-19T12:58:09+01:00] INFO: Processing directory[C:/Temp/B] action create (t
est::default line 7)
[2015-10-19T12:58:09+01:00] INFO: directory[C:/Temp/B] created directory C:/Temp/B

    - create new directory C:/Temp/B
  * link[C:/Temp/A] action create[2015-10-19T12:58:09+01:00] INFO: Processing link[C:/Temp/A] action create (test::defau
lt line 10)
[2015-10-19T12:58:09+01:00] INFO: link[C:/Temp/A] created

    - create symlink at C:/Temp/A to C:/Temp/B
  * link[C:/Temp/A] action delete[2015-10-19T12:58:09+01:00] INFO: Processing link[C:/Temp/A] action delete (test::defau
lt line 13)


    ================================================================================
    Error executing action `delete` on resource 'link[C:/Temp/A]'
    ================================================================================

    Errno::EACCES
    -------------
    Permission denied - C:/Temp/A

    Resource Declaration:
    ---------------------
    # In c:/chef/local-mode-cache/cache/cookbooks/test/recipes/default.rb

     13: link "C:/Temp/A" do
     14:   action :delete
     15: end

    Compiled Resource:
    ------------------
    # Declared in c:/chef/local-mode-cache/cache/cookbooks/test/recipes/default.rb:13:in `from_file'

    link("C:/Temp/A") do
      action [:delete]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      to "C:/Temp/B"
      link_type :symbolic
      target_file "C:/Temp/A"
      declared_type :link
      cookbook_name "test"
      recipe_name "default"
    end

[2015-10-19T12:58:09+01:00] INFO: Running queued delayed notifications before re-raising exception

Running handlers:
[2015-10-19T12:58:09+01:00] ERROR: Running exception handlers
Running handlers complete
[2015-10-19T12:58:09+01:00] ERROR: Exception handlers complete
Chef Client failed. 2 resources updated in 21 seconds
[2015-10-19T12:58:09+01:00] FATAL: Stacktrace dumped to c:/chef/local-mode-cache/cache/chef-stacktrace.out
[2015-10-19T12:58:09+01:00] FATAL: Errno::EACCES: link[C:/Temp/A] (test::default line 13) had an error: Errno::EACCES: P
ermission denied - C:/Temp/A
```

Using irb, 
```
irb(main):001:0> File.delete('C:/Temp/A')
Errno::EACCES: Permission denied - C:/Temp/A
        from (irb):1:in `delete'
        from (irb):1
        from C:/opscode/chef/embedded/bin/irb:12:in `<main>'
```
whereas 
```
irb(main):002:0> Dir.delete('C:/Temp/A')
=> 0
```

File.delete is called here:
https://github.com/chef/chef/blob/master/lib/chef/provider/link.rb#L142-L149

I guess we need to check if the target is a directory, and call Dir.delete instead? (but only on Windows?)

**Comment by thommay** [comment]
ping @chef/client-windows

**Comment by mwrock** [comment]
I'd bed that `Dir.delete` also works on linux. So we could have a single code path apply to both.

**Comment by delivery-test** [comment]
@mwrock `Dir.delete` does not work on Linux for any kind of link. It works fine with `File.delete`. I have verified this and fixed [here](https://github.com/chef/chef/pull/5234).

**Comment by btm** [comment]
This is fixed by #5234 and will be in the next release.
